u5-rest-apis/features/book-controller.js

```diff
@@ -0,0 +1,251 @@
+      result: 'error',
+      payload: [],
+      count: 0
+    })
+  }
+}
+
+const readBook = async (req, res) => {
+  try {
+    // 1. Inputs
+    const isbn13 = req.params.isbn13
+
+    // 2. Query
+    const query = db.collection('books').doc(isbn13).get()
+
+    // 3. Response
+    const snapshot = await query
+    if (!snapshot.exists) { return res.status(404).json({ result: 'not found' }) }
+
+    const { title, authors, description, pages } = snapshot.data()
+    const payload = { title, authors, description, pages }
+    res.json({ result: 'ok', payload })
+  } catch (err) {
+    console.error(err)
+    res.status(500).json({ result: 'error' })
+  }
+}
+
+const createBook = async (req, res) => {
+  try {
+    // 1. Inputs
+    const { isbn13, title, authors, description, pages } = req.body
+    const book = { isbn13, title, authors, description, pages: parseInt(pages) || 0 }
+
+    // 2. Query
+    const query = db.collection('books').doc(isbn13).set(book, { merge: true })
+
+    // 3. Response
+    await query
+    res.sendStatus(201)
+  } catch (err) {
+    console.error(err)
+    res.sendStatus(500)
+  }
+}
+
+const replaceBook = async (req, res) => {
+  try {
+    // 1. Inputs
+    const isbn13 = req.params.isbn13
+    const { title, authors, description, pages } = req.body
+    const book = { isbn13, title, authors, description, pages: parseInt(pages) }
+
+    // 2. Query
+    const query = db.collection('books').doc(isbn13).set(book) // no merge
+
+    // 3. Response
+    await query
+    res.sendStatus(200)
+  } catch (err) {
+    console.error(err)
+    res.sendStatus(500)
+  }
+}
+const updateBook = async (req, res) => {
+  try {
+    // 1. Inputs
+    const isbn13 = req.params.isbn13
+    const { title } = req.body
+    const book = { title }
+
+    // 2. Query
+    const query = db.collection('books').doc(isbn13).set(book, { merge: true })
+
+    // 3. Response
+    await query
+    res.sendStatus(200)
+  } catch (err) {
+    console.error(err)
+    res.sendStatus(500)
+  }
+}
+
+const deleteBook = async (req, res) => {
+  try {
+    // 1. Inputs
+    const isbn13 = req.params.isbn13
+
+    // 2. Query
+    const query = db.collection('books').doc(isbn13).delete()
+
+    // 3. Response
+    await query
+    res.sendStatus(200)
+  } catch (err) {
+    console.error(err)
+    res.sendStatus(500)
+  }
+}
+
+/// ///////////////////////////// Covid \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\
+
+const readCovidRecords = async (req, res) => {
+  try {
+    // 1. Inputs
+    // none
+
+    // 2. Query
+    const query = db.collection('covid-latest').get()
+
+    // 3. Response
+    const payload = (await query)
+      .docs
+      .map(doc => doc.data())
+      .map(({ date, stateId, stateName, cases, casesNew, vaccineOne, vaccineOnePercent, vaccineComplete, vaccineCompletePercent }) =>
+        ({ date: date.toMillis(), stateId, stateName, cases, casesNew, vaccineOne, vaccineOnePercent, vaccineComplete, vaccineCompletePercent }))
```

Comment:
**โค้ดนี้สร้างอ็อบเจ็กต์ใหม่ที่มีเฉพาะฟิลด์ที่เราต้องการแชร์** ไม่ใช่เรื่องง่าย น่าเสียดายที่ฉันไม่พบวิธีแก้ปัญหาที่อ่านง่ายกว่านี้ ​นี่คือความคิดที่ดีที่สุด 3 ข้อของฉัน (พวกเขาทั้งหมดมีปัญหา)

**This code creates new objects containing only the fields we want to share.** I think it's confusing. Unfortunately I can't find a more readable solution. These are my 3 best ideas (they all have problems):

## หลายวัตถุ:

```js
// (1) เพียงแค่สร้างวัตถุ (JS พื้นฐาน)
.map(
  data => ({ date: data.date.toMillis(), stateId: data.stateId, stateName: data.stateName, ETC })
)

// (2) Destructure + short-form
.map(
  ({ date, stateId, stateName, ETC }) => ({ date: date.toMillis(), stateId, stateName, ETC })
)

// (3) Lodash (ต้องติดตั้งไลบรารี npm)
.map(
  data => pick(data, ['stateId', 'stateName', ETC])
  // ...หรือด้วยการแปลงวันที่:
  data => { ...pick(data, ['stateId', 'stateName', ETC]), date: data.date.toMillis() }
)
```

## วัตถุเดียว:

```js
// (1) เพียงแค่สร้างวัตถุ (JS พื้นฐาน)
// +: เข้าใจง่าย
// -: เขียนชื่อ field สองครั้ง; เขียน `data` หลายครั้ง
const myObject = { 
  date: Timestamp.fromMillis(req.body.date),
  stateId: req.body.stateId,
  stateName: req.body.stateName, 
  ETC
}

// (2) Destructure + short-form
// +: สั้นลงเล็กน้อย
// -: อ่านง่ายน้อยลง; ง่ายที่จะลืมเพิ่ม field ทางด้านขวา
const { date, stateId, stateName, ETC } = req.body
const myObject = { date: Timestamp.fromMillis(date), stateId, stateName, ETC }

// (3) Lodash (ต้องติดตั้งไลบรารี npm)
// +: เขียนช่องเพียงครั้งเดียว
// -: ไม่รองรับการแปลง field; ชื่อ field เป็นสตริง (VSCode ตรวจไม่พบการพิมพ์ผิด)
const myObject = pick(req.body, ['stateId', 'stateName', ETC])
// ...หรือด้วยการแปลงวันที่:
const myObject = { ...pick(req.body, ['stateId', 'stateName', ETC]), date: Timestamp.fromMillis(req.body.date) }
```